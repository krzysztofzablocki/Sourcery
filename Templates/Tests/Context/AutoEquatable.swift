//
//  AutoCases.swift
//  Templates
//
//  Created by Anton Domashnev on 03.05.17.
//  Copyright Â© 2017 Pixle. All rights reserved.
//

import Foundation

protocol AutoEquatable {}

protocol Parent {
    var name: String { get }
}

/// General protocol
protocol AutoEquatableProtocol: AutoEquatable {
    var width: Double { get }
    var height: Double { get}
    static var name: String { get }
}

/// General enum
enum AutoEquatableEnum: AutoEquatable {
    case one
    case two(first: String, second: String)
    case three(bar: Int)

    func allValue() -> [AutoEquatableEnum] {
        return [.one, .two(first: "a", second: "b"), .three(bar: 42)]
    }
}

/// Sourcery should not generate a default case for enum with only one case
enum AutoEquatableEnumWithOneCase: AutoEquatable {
    case one
}

/// Sourcery should generate correct code for struct
struct AutoEquatableStruct: AutoEquatable {
    // Private Constants
    private let laptopModel: String

    // Fileprivate Constants
    fileprivate let phoneModel: String

    // Internal Constants
    let firstName: String

    // Public Constans
    public let lastName: String

    init(firstName: String, lastName: String, parents: [Parent], laptopModel: String, phoneModel: String) {
        self.firstName = firstName
        self.lastName = lastName
        self.parents = parents
        self.laptopModel = laptopModel
        self.phoneModel = phoneModel
    }

    // Arrays
    // sourcery: arrayEquality
    let parents: [Parent]

    // Variable
    var moneyInThePocket: Double = 0

    // Optional variable
    var friends: [String]?

    /// Forced unwrapped variable
    var age: Int!

    // Void method
    func walk() {
        print("I'm going")
    }

    // Method with return value
    func greeting(for name: String) -> String {
        return "Hi \(name)"
    }

    // Method with optional return value
    func books(sharedWith name: String) -> String? {
        return nil
    }
}

/// It should generate correct code for general class
class AutoEquatableClass: AutoEquatable {
    // Private Constants
    private let laptopModel: String

    // Fileprivate Constants
    fileprivate let phoneModel: String

    // Internal Constants
    let firstName: String

    // Public Constans
    public let lastName: String

    init(firstName: String, lastName: String, parents: [Parent], laptopModel: String, phoneModel: String) {
        self.firstName = firstName
        self.lastName = lastName
        self.parents = parents
        self.laptopModel = laptopModel
        self.phoneModel = phoneModel
    }

    // Arrays
    // sourcery: arrayEquality
    let parents: [Parent]

    /// Forced unwrapped variable
    var age: Int!

    // Variable
    var moneyInThePocket: Double = 0

    // Optional variable
    var friends: [String]?

    // Void method
    func walk() {
        print("I'm going")
    }

    // Method with return value
    func greeting(for name: String) -> String {
        return "Hi \(name)"
    }

    // Method with optional return value
    func books(sharedWith name: String) -> String? {
        return nil
    }
}

/// Sourcery doesn't support inheritance for AutoEqualtable 
class AutoEquatableClassInherited: AutoEquatableClass {
    // Optional constants
    let middleName: String?

    init(middleName: String?) {
        self.middleName = middleName
        super.init(firstName: "", lastName: "", parents: [], laptopModel: "", phoneModel: "")
    }
}

/// Should not add Equatable conformance
class AutoEquatableNSObject: NSObject, AutoEquatable {
    let firstName: String

    init(firstName: String) {
        self.firstName = firstName
    }
}
