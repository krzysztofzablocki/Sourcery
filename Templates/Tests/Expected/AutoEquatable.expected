// swiftlint:disable file_length
fileprivate func compareOptionals<T>(lhs: T?, rhs: T?, compare: (_ lhs: T, _ rhs: T) -> Bool) -> Bool {
    switch (lhs, rhs) {
    case let (lValue?, rValue?):
        return compare(lValue, rValue)
    case (nil, nil):
        return true
    default:
        return false
    }
}

fileprivate func compareArrays<T>(lhs: [T], rhs: [T], compare: (_ lhs: T, _ rhs: T) -> Bool) -> Bool {
    guard lhs.count == rhs.count else { return false }
    for (idx, lhsItem) in lhs.enumerated() {
        guard compare(lhsItem, rhs[idx]) else { return false }
    }

    return true
}


// MARK: - AutoEquatable for classes, protocols, structs
// MARK: - AutoEquatableClass AutoEquatable
extension AutoEquatableClass: Equatable {}
internal func == (lhs: AutoEquatableClass, rhs: AutoEquatableClass) -> Bool {
    guard lhs.firstName == rhs.firstName else { return false }
    guard lhs.lastName == rhs.lastName else { return false }
    guard compareArrays(lhs: lhs.parents, rhs: rhs.parents, compare: ==) else { return false }
    guard compareOptionals(lhs: lhs.age, rhs: rhs.age, compare: ==) else { return false }
    guard lhs.moneyInThePocket == rhs.moneyInThePocket else { return false }
    guard compareOptionals(lhs: lhs.friends, rhs: rhs.friends, compare: ==) else { return false }
    return true
}
// MARK: - AutoEquatableClassInherited AutoEquatable
extension AutoEquatableClassInherited: Equatable {}
THIS WONT COMPILE, WE DONT SUPPORT INHERITANCE for AutoEquatable
internal func == (lhs: AutoEquatableClassInherited, rhs: AutoEquatableClassInherited) -> Bool {
    guard compareOptionals(lhs: lhs.middleName, rhs: rhs.middleName, compare: ==) else { return false }
    return true
}
// MARK: - AutoEquatableNSObject AutoEquatable
internal func == (lhs: AutoEquatableNSObject, rhs: AutoEquatableNSObject) -> Bool {
    guard lhs.firstName == rhs.firstName else { return false }
    return true
}
// MARK: - AutoEquatableProtocol AutoEquatable
internal func == (lhs: AutoEquatableProtocol, rhs: AutoEquatableProtocol) -> Bool {
    guard lhs.width == rhs.width else { return false }
    guard lhs.height == rhs.height else { return false }
    guard lhs.name == rhs.name else { return false }
    return true
}
// MARK: - AutoEquatableStruct AutoEquatable
extension AutoEquatableStruct: Equatable {}
internal func == (lhs: AutoEquatableStruct, rhs: AutoEquatableStruct) -> Bool {
    guard lhs.firstName == rhs.firstName else { return false }
    guard lhs.lastName == rhs.lastName else { return false }
    guard compareArrays(lhs: lhs.parents, rhs: rhs.parents, compare: ==) else { return false }
    guard lhs.moneyInThePocket == rhs.moneyInThePocket else { return false }
    guard compareOptionals(lhs: lhs.friends, rhs: rhs.friends, compare: ==) else { return false }
    guard compareOptionals(lhs: lhs.age, rhs: rhs.age, compare: ==) else { return false }
    return true
}

// MARK: - AutoEquatable for Enums
// MARK: - AutoEquatableEnum AutoEquatable
extension AutoEquatableEnum: Equatable {}
internal func == (lhs: AutoEquatableEnum, rhs: AutoEquatableEnum) -> Bool {
    switch (lhs, rhs) {
    case (.one, .one):
        return true
    case (.two(let lhs), .two(let rhs)):
        if lhs.first != rhs.first { return false }
        if lhs.second != rhs.second { return false }
        return true
    case (.three(let lhs), .three(let rhs)):
        return lhs == rhs
    default: return false
    }
}
// MARK: - AutoEquatableEnumWithOneCase AutoEquatable
extension AutoEquatableEnumWithOneCase: Equatable {}
internal func == (lhs: AutoEquatableEnumWithOneCase, rhs: AutoEquatableEnumWithOneCase) -> Bool {
    switch (lhs, rhs) {
    case (.one, .one):
        return true
    }
}
